### ISSUE TITLE
Document given CanEqual on companion objects

### ISSUE THREAD

**Issue by amitport** [lampepfl/dotty#12716]
## Compiler version

`scala 3.0`

## Minimized example

```Scala
import scala.language.strictEquality

case class Suit(value: String)

object Suit:
  given CanEqual[Suit, Suit] = CanEqual.derived

@main def main =
  println(Suit("X") == Suit("X"))
```

## Output
```scala
true
```

## Expectation

According to 
https://github.com/lampepfl/dotty/blob/a9a7e7042408dd0de6cebdddc5fc23a6e33d26be/docs/docs/reference/contextual/multiversal-equality.md#precise-rules-for-equality-checking
objects can be compared using `==` under `strictEquality` if there is a `given` of type `CanEqual[T, U]`.

In the example above it's not clear (to me) and I can't find documentation as to why the given CanEqual in the `companion object` is available in this case. AFAICT, it is in general not available as given at the main scope without an import. 

Thank you

**Comment by odersky** [comment]
It is in the implicit scope. The same rules as for all other givens apply.

**Comment by amitport** [comment]
@odersky sorry, it is really not clear to me what happens in the case above. Maybe this is a documentation issue with http://dotty.epfl.ch/docs/reference/contextual/using-clauses.html or http://dotty.epfl.ch/docs/reference/contextual/givens.html but I strongly believe is is an issue. I know it is easy to say RTFM, but please consider that I have although I admit I may have missed something, I did have it another pass and would appreciate at least some reference that does explain this.

**Comment by som-snytt** [comment]
@amitport it works like this
```
scala> trait T[A]
// defined trait T

scala> class C ; object C { given T[C] = null }
// defined class C
// defined object C

scala> summon[T[C]]
val res0: T[C] = null
```

It's explained as the associated parts in https://scala-lang.org/files/archive/spec/2.13/07-implicits.html#implicit-parameters, where the type arg `C` in `T[C]` means the companion of `C` is in implicit scope.

**Comment by amitport** [comment]
@som-snytt thank you! this is indeed stated clearly in your link. 

Though I still believe this should be a documentation issue regarding the relevant parts in scala 3 book and dotty's documentation. 

Yes, if you go backward you can find notes about the implicit scope. That's probably why when I wrote down my code, I automatically used the companion object (I used scala around 2.8/10 and now trying to evaluate scala 3), but, when I wanted to understand exactly what has happened I couldn't find the relevant documentation in the context of scala 3.

So, in short, thank you for your responses, as far as `CanEqual` this should probably be closed, but IMHO a more general issue for documenting the implicit scope resolution in scala 3 documentation should be opened.

Thank you, again

**Comment by som-snytt** [comment]
@amitport The amended doc is at https://dotty.epfl.ch/docs/reference/changed-features/implicit-resolution.html

The relevant section is Nr 3, which I had previously skipped because of tldr;  The terminology is revised but the gist is similar.

There is a TODO on the page to refine it.

I agree that if I've read something in the reference docs, but I need to find it again because I haven't had enough opportunity to use Scala 3, it can be hard to find what I need. Sometimes I grep the repo.

I don't know how this topic is covered in introductory materials.

**Comment by amitport** [comment]
Thanks, hopefully, at some point scala3 language guides would contain all the language semantics (without referring to scala2 language specification + migration guide). Implicit resolution is pretty much lacking in scala2 documentation also (it's in the specification, but to me, it seems it should be right there where, implicit/using clauses are introduced)
